Reproducible Failure in IndexShardTests.testRestoreLocalHistoryFromTranslogOnPromotion

Reproduces 100% of the time for me locally on master:

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.index.shard.IndexShardTests.testRestoreLocalHistoryFromTranslogOnPromotion" -Dtests.seed=8F7E1836ED070021 -Dtests.security.manager=true -Dtests.locale=nl-BE -Dtests.timezone=Brazil/West -Druntime.java=14

org.elasticsearch.index.shard.IndexShardTests > testRestoreLocalHistoryFromTranslogOnPromotion FAILED
    java.lang.AssertionError: 
    Expected: <12L>
         but: was <10L>
        at __randomizedtesting.SeedInfo.seed([8F7E1836ED070021:3FB466971CC634D8]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.index.shard.IndexShardTests.testRestoreLocalHistoryFromTranslogOnPromotion(IndexShardTests.java:1206)


Suite: Test class org.elasticsearch.index.shard.IndexShardTests
  1> [2020-06-16T06:07:53,046][INFO ][o.e.i.s.IndexShardTests  ] [testRestoreLocalHistoryFromTranslogOnPromotion] before test
  1> [2020-06-16T06:07:53,906][INFO ][o.e.i.s.IndexShard       ] [org.elasticsearch.index.shard.IndexShardTests] [index][0] detected new primary with primary term [7], global checkpoint [12], max_seq_no [12]
  1> [2020-06-16T06:07:53,911][INFO ][o.e.i.s.IndexShardTests  ] [testRestoreLocalHistoryFromTranslogOnPromotion] after test
  2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.index.shard.IndexShardTests.testRestoreLocalHistoryFromTranslogOnPromotion" -Dtests.seed=8F7E1836ED070021 -Dtests.security.manager=true -Dtests.locale=nl-BE -Dtests.timezone=Brazil/West -Druntime.java=14
  2> java.lang.AssertionError: 
    Expected: <12L>
         but: was <10L>
        at __randomizedtesting.SeedInfo.seed([8F7E1836ED070021:3FB466971CC634D8]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.index.shard.IndexShardTests.testRestoreLocalHistoryFromTranslogOnPromotion(IndexShardTests.java:1206)
  2> NOTE: leaving temporary files on disk at: /home/brownbear/src/elasticsearch/server/build/testrun/test/temp/org.elasticsearch.index.shard.IndexShardTests_8F7E1836ED070021-003
  2> NOTE: test params are: codec=Asserting(Lucene86): {}, docValues:{}, maxPointsInLeafNode=1049, maxMBSortInHeap=5.637758798428584, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@251926d5), locale=nl-BE, timezone=Brazil/West
  2> NOTE: Linux 5.4.0-37-generic amd64/AdoptOpenJDK 14.0.1 (64-bit)/cpus=16,threads=1,free=491558296,total=536870912
  2> NOTE: All tests run in this JVM: [IndexShardTests]

1 test completed, 1 failed
```

--- a/server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
+++ b/server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
@@ -1182,13 +1182,13 @@ public class IndexShardTests extends IndexShardTestCase {
         indexShard.updateGlobalCheckpointOnReplica(globalCheckpointOnReplica, "test");
 
         final long globalCheckpoint = randomLongBetween(UNASSIGNED_SEQ_NO, indexShard.getLocalCheckpoint());
-        final long maxSeqNoOfUpdatesOrDeletes = randomLongBetween(SequenceNumbers.NO_OPS_PERFORMED, maxSeqNo);
+        final long maxSeqNoOfUpdatesOrDeletesBeforeRollback = indexShard.getMaxSeqNoOfUpdatesOrDeletes();
         final Set<String> docsBeforeRollback = getShardDocUIDs(indexShard);
         final CountDownLatch latch = new CountDownLatch(1);
         randomReplicaOperationPermitAcquisition(indexShard,
                 indexShard.getPendingPrimaryTerm() + 1,
                 globalCheckpoint,
-                maxSeqNoOfUpdatesOrDeletes,
+                randomLongBetween(SequenceNumbers.NO_OPS_PERFORMED, maxSeqNo),
                 new ActionListener<Releasable>() {
                     @Override
                     public void onResponse(Releasable releasable) {
@@ -1203,7 +1203,11 @@ public class IndexShardTests extends IndexShardTestCase {
                 }, "");
 
         latch.await();
-        assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNo));
+        if (globalCheckpoint < maxSeqNo) {
+            assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNo));
+        } else {
+            assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNoOfUpdatesOrDeletesBeforeRollback));
+        }
         final ShardRouting newRouting = indexShard.routingEntry().moveActiveReplicaToPrimary();
         final CountDownLatch resyncLatch = new CountDownLatch(1);
         indexShard.updateShardState(
@@ -1217,7 +1221,11 @@ public class IndexShardTests extends IndexShardTestCase {
         assertThat(indexShard.getLocalCheckpoint(), equalTo(maxSeqNo));
         assertThat(indexShard.seqNoStats().getMaxSeqNo(), equalTo(maxSeqNo));
         assertThat(getShardDocUIDs(indexShard), equalTo(docsBeforeRollback));
-        assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNo));
+        if (globalCheckpoint < maxSeqNo) {
+            assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNo));
+        } else {
+            assertThat(indexShard.getMaxSeqNoOfUpdatesOrDeletes(), equalTo(maxSeqNoOfUpdatesOrDeletesBeforeRollback));
+        }
         closeShard(indexShard, false);
     }
 

